src/integrators/wrapper_integrator.hpp

```diff
@@ -156,6 +157,72 @@ class MixedBilinearToNonlinearFormIntegrator : public mfem::NonlinearFormIntegra
   std::shared_ptr<mfem::ParFiniteElementSpace> trial_fes_;
 };
 
+/**
+ * @brief A class to convert NonlinearFormIntegrator to one where the input parameter undergoes a change of variables
+ */
+class TransformedNonlinearFormIntegrator : public mfem::NonlinearFormIntegrator {
+public:
+  /// alias for prototype of the residual_func
+  using transformed_func = std::shared_ptr<mfem::Vector>(const mfem::FiniteElement& el, mfem::ElementTransformation& Tr,
+                                                         const mfem::Vector&);
+
+  /// alias for prototype of the gradient of residual_func
+  using transformed_gradient_func = std::shared_ptr<mfem::DenseMatrix>(const mfem::FiniteElement&   el,
+                                                                       mfem::ElementTransformation& Tr,
+                                                                       const mfem::DenseMatrix&);
+
+  /**
+   * @brief Recasts, A(u(x)) = F as R(u(x)) = A(u(x)) - F = R(x)
+   *
+   * @param[in] R A BilinearFormIntegrator
+   * @param[in] transformed_input A function that performs a change of variables to what R expects
+   * @param[in] transformed_grad_output A function that performs a change of variables for the gradient
+   */
+  explicit TransformedNonlinearFormIntegrator(std::shared_ptr<mfem::NonlinearFormIntegrator> R,
+                                              std::function<transformed_func>                transformed_input,
+                                              std::function<transformed_gradient_func>       transformed_grad_output);
+
+  /**
+   * @brief Compute the residual vector with input, transformed_function(x)
+   * @param[in] el The finite element for local integration
+   * @param[in] Tr The local FE transformation
+   * @param[in] elfun The state to evaluate the residual
+   * @param[out] elvect The output residual
+   */
+  virtual void AssembleElementVector(const mfem::FiniteElement& el, mfem::ElementTransformation& Tr,
+                                     const mfem::Vector& elfun, mfem::Vector& elvect);
+
+  /**
+   * @brief Compute the tangent matrix with input, transformed_function(x)
+   *
+   * @param[in] el The finite element for local integration
+   * @param[in] Tr The local FE transformation
+   * @param[in] elfun The state to evaluate the residual
+   * @param[out] elmat elvect The output gradient
+   */
+  virtual void AssembleElementGrad(const mfem::FiniteElement& el, mfem::ElementTransformation& Tr,
+                                   const mfem::Vector& elfun, mfem::DenseMatrix& elmat);
+
+private:
+  /**
+   * @brief The NonlinearFormIntegrator form to wrap
+   *
+   */
+  std::shared_ptr<mfem::NonlinearFormIntegrator> R_;
```

Comment:
Could this be a `unique_ptr`?

Reply:
Great question. 

My intention was it for to be shared, so someone can create a base mfem integrator and then one can modify the behavior of this integrator by using the `Transformed*` functions. For example in some of the time stepping algorithms, we might have different `Transformed*` functions for explicit and implicit time integration, but they are based on the same initial integrator.

Currently when the `mfem::NonlinearForm` goes out of scope, it will automatically delete all of it's integrators. In this case the original `NonlinearFormIntegrator` which may be defined and used elsewhere also won't be deleted. This is nice as we can modify different numerical behaviors by modifying the base integrator for different problems and potentially in comparing different time integration methods.

I'm open to other ways of implementing this.